### ISSUE TITLE
Feature Request: Force Build (for development workflows)

### ISSUE THREAD

**Issue by nigelgbanks** [6si/shipwright#42]
Currently shipwright will build all the images in a given repository for the current commit. Which is great if you do not plan on making changes within the repository, as subsequent calls to shipwright will not trigger a rebuild if the files are changed. 

Often though I have several Dockerfiles in a single repository and I'm actively working on a number of them, making changes and re-building (but not committing). 

It would be ideal for me, if I didn't have to think about the build order and I could just have shipwright rebuild the changed images as need be.

Perhaps this could be achieved by adding an additional flag, say ```-b``` to always trigger a build. Additionally we could also expose ```--no-cache`` from the Docker client as well, it we want Docker to build without the cache.

**Comment by nigelgbanks** [comment]
I'd be happy to collaborate with the implementation of this feature.

**Comment by srobertson** [comment]
Yes, this a huge pain point for us internally. The main issue is shipwright only looks at the commit log to determine if something needs to be rebuilt. It doesn't care about uncommitted changes.

Not a fan of the always build approach unless we can specify it by image. If you have a big tree that could take forever.

The ideal approach would use git/pygit to determine if there are uncommitted changes present and only rebuild if the md5s have changed.

Eyeballing the code it looks like a good place to wire that in would be in shipwright/__init__.py. 

Basically modify the Shipwright.targets() method to calculate target.current_rel, target.last_built_rel differently when in development mode.

**Comment by nigelgbanks** [comment]
I like the idea of only building if their are uncommitted changes, it's a lot more intuitive and wouldn't require a change to the CLI.

As for the tag I think "uncommitted" is pretty straight forward, perhaps suffixed with the last commit hash so something like "uncommitted-a55516346947" or prefixed like so "a55516346947-uncommitted", so one could switch between branches and commits and the tag's wouldn't collide.

**Comment by srobertson** [comment]
Wonder if it's worth looking into git's algorithm for generating the hash commits based on the set of changes to be made. In theory all the changes we're about to apply would consistently hash to the next commit. Which means if you get it right while developing and commit based on those changes then you won't have to build a second time.

Could be cool, could be to much effort.

**Comment by nigelgbanks** [comment]
I believe the git commit hash is also based on the git commit message (for
example amending the commit message changes the hash), so I don't think the
hashes wouldn't line.

When the build is triggered a second time when the commit is made, it could
at least use the Docker image cache so the builds shouldn't take too long
either way.

On Tue, Feb 23, 2016 at 11:42 PM, Scott Robertson <notifications@github.com>
wrote:

> Wonder if it's worth looking into git's algorithm for generating the hash
> commits based on the set of changes to be made. In theory all the changes
> we're about to apply would consistently hash to the next commit. Which
> means if you get it right while developing and commit based on those
> changes then you won't have to build a second time.
>
> Could be cool, could be to much effort.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/6si/shipwright/issues/42#issuecomment-187947124>.
>



-- 
Nigel Banks
nigel.g.banks@gmail.com

**Comment by srobertson** [comment]
Agreed

**Comment by graingert** [comment]
@nigelgbanks the commit message and the timestamp

**Comment by nigelgbanks** [comment]
Ok so if I could summarize this is how we'd like it to function (please comment with other idea's or if there might be a problem with the implementation).

1. So as @srobertson said we can use git/pygit to detect changes in a folder by folder bases (those that contain Dockerfiles). 

2. We would then build those that had changed and tag them with the commit has plus the suffix **uncommitted**, like so: **a55516346947-uncommitted** (these would also get the branch tag, and the latest tag).

3. Any images dependant on a changed image would also be triggered to build, based on the tag from step 2. They will also receive the **a55516346947-uncommitted** tag along with the branch tag, and the latest tag. 

4. The purge command would be updated to remove the **uncommitted** images as well.

I think this is good for a first pass, perhaps later we could also append a hash to **a55516346947-uncommitted** that captures the changes in the directory so builds would only be triggered if the folder change since the last call to shipwright.

**Comment by srobertson** [comment]
I think this will work. It's a great first pass.

Is the value of tagging it `a55516346947-uncommitted` so that we use the local docker registry to track the last built? 

I'm suspecting from a usability perspective the dev will just want the image tagged `latest` every time and will want us to have no more than one uncommitted tag per image at a time. I tend to run out of space when shipwright tags to many things.

That being said, I think we get `latest` for free and we can worry about uncommited cleanup as a second pass.

**Comment by jalaziz** [comment]
Minor comment, but I'd recommend using `-dirty` instead of `-uncommitted`. Or better yet, use `git describe` and potentially allow the user to specify the `mark` for the `--dirty` flag.

**Comment by nigelgbanks** [comment]
I was thinking tagging with **uncommitted** or **dirty** as @jalaziz suggests; so that we could switch branches at will and run both builds. 

Though I think as @srobertson points out; under the typical use case, the developer would probably only want the images to be tagged **latest**. As normally when developing the user will work on current branch exclusively and adding too many tags makes things more complicated for the end user as well as consumes more space.

**Comment by dunk** [comment]
I'd love this feature too. Can't really use shipwright without it.

I'll throw in too on what is sure to be a very bikesheddy issue soon. "uncommitted" as a tag still presupposes that the user's thought process revolves around the commit state of their code. But I'd wager that while developing something more intuitive like "dev", "development", or similar would be a better fit. You're not going to push these containers to a local registry anyways, because you'd always check in your code anyways before sharing with another developer.

**Comment by jamespic** [comment]
One approach that might solve this problem indirectly would be to add a `shipwright clean` command, that untags/deletes images that have been built previously. You could then run a dirty build with `shipwright clean && shipwright build [-t dirty]`. I reckon I could implement this with relatively few changes to existing code.

I'm happy to do the work, if the broad approach is satisfactory.

**Comment by graingert** [comment]
@jamespic I'd rather discourage cleaning because it should always be possible to create a hash that represents the changes

**Comment by graingert** [comment]
@jamespic it should be possible to create a hash of the diff information from:

```repo.index.diff(None)```

Go through the list of diffs, creating a list of (source_name, a_path, b_path, a_blob, b_blob), then sort generate a hash out of that list.

If you want to be really smart you should filter the list by elements with a_path and b_path that share a prefix with any path in "paths" at https://github.com/6si/shipwright/blob/c086fa4a73f350efe1a78a73e397273ea7f26dfa/shipwright/source_control.py#L87

**Comment by graingert** [comment]
then just append that hash onto "hexsha"

**Comment by graingert** [comment]
then it's a case of, do you want to track uncommited files?

**Comment by jamespic** [comment]
@graingert That sounds simple enough. Would the intention be that users specifically add a "dirty" flag, or that it "just works" (e.g, generating a dirty image whenever there are staged but uncommitted changes)?

**Comment by graingert** [comment]
@jamespic if there are any diffed files shipwright should crash with an error, unless "--dirty" is added.

When "--dirty" is added you get builds tags like `f"{hexsha(_last_commit(repo, paths))}-{_hash_diff(repo, paths)[:12]}"`

**Comment by jamespic** [comment]
Cool. That sounds doable.

I suspect you're thinking a couple of problems ahead of me, but I don't quite grok what you're saying about filtering shared prefix paths. Care to elaborate?

**Comment by graingert** [comment]
assuming you have a function:
```
def _diffs_for_repo(repo): -> [(a_path, b_path, a_blob, b_blob), ...]
```
you should do something like:
```python
# psuedocode
def _hash_diff(repo, paths):
    diffs = _diffs_for_repo(repo)
    changed_in_path = []
    for diff in diffs:
        (a_path, b_path, *_) = diff
        if a_path.startswith(paths) or b_path.startswith(paths):
            changed_in_path.append(diff)
```

**Comment by graingert** [comment]
"paths" in this case is a list of paths containing the image source to be built and those of its dependencies, so you only rebuild an image if it or its parents have changed

**Comment by graingert** [comment]
Just merged this feature, thanks @jamespic!
